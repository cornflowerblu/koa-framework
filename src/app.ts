import Koa from 'koa'
import Router from 'koa-router'
import getRoute from './routes'

const app = new Koa()
const router = new Router()

router.get(getRoute('home'), (ctx: Koa.Context) => {
  ctx.body = 'Hello World'
})

app.use(router.routes())

export default app
