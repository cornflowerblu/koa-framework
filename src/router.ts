import Koa from 'koa'
import Router from 'koa-router'
import invariant from 'tiny-invariant'
import { createNewUser, getManyUsers } from './controllers/user'
import {
  getProtectedHomepage,
  getUnprotectedHomepage,
} from './controllers/homepage'

type Routes = {
  name: string
  path: string
}

// Routing function that lets us use custom objects with human-friendly names and anything else we may want down the line
export const getRoute = (name: string): string => {
  let routeName = routes.find((el) => el.name === name)
  invariant(routeName)
  return routeName?.path
}

export const router = new Router()

// The list of routes. Name and Path are required. 
const routes: Array<Routes> = [
  {
    name: 'home page protected',
    path: '/',
  },
  {
    name: 'landing page un-protected',
    path: '/page',
  },
  {
    name: 'create new user',
    path: '/user/create',
  },
  {
    name: 'get all users',
    path: '/users',
  },
]

// Look how easy these are to read!
router.get(getRoute('home page protected'), (ctx: Koa.Context) =>
  getProtectedHomepage(ctx)
)

router.get(getRoute('landing page un-protected'), (ctx: Koa.Context) =>
  getUnprotectedHomepage(ctx)
)

router.post(getRoute('create new user'), (ctx: Koa.Context) =>
  createNewUser(ctx)
)

router.get(getRoute('get all users'), (ctx: Koa.Context) => getManyUsers(ctx))

export default routes
