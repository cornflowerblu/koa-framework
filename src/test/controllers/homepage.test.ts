import request from 'supertest'
import app from '../../app'
import { getRoute } from '../../router'

test('Unauthorized page throws 403', async () => {
  const url = getRoute('home page protected')
  const response = await request(app.callback()).get(url)
  expect(response.status).toBe(403)
  expect(response.text).toMatchSnapshot()
})

test('Un-protected page returns a response', async () => {
  const url = getRoute('landing page un-protected')
  const response = await request(app.callback()).get(url)
  expect(response.status).toBe(200)
  expect(response.text).toMatchSnapshot()
})
