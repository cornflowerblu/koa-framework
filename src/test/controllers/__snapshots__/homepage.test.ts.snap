// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Un-protected page returns a response 1`] = `"Hi!"`;

exports[`Unauthorized page throws 403 1`] = `
"{
  \\"statusCode\\": 403,
  \\"error\\": \\"Forbidden\\",
  \\"message\\": \\"Forbidden\\"
}"
`;
